### ISSUE TITLE
Attributes aren't visible inside custom evaluation function in CV

### ISSUE THREAD

**Issue by BlindApe** [Microsoft/LightGBM#581]
Attributes are a tricky method to define custom evaluation metrics (evaluation by groups, evaluation only in some interested subset...)

xgboost works fine with them but with lightgbm they aren't visible inside the custom metric.
I suspect this could be a problem with slice method.

In xgboost:
```
slice.xgb.DMatrix <- function(object, idxset, ...) {
  if (!inherits(object, "xgb.DMatrix")) {
    stop("object must be xgb.DMatrix")
  }
  ret <- .Call(XGDMatrixSliceDMatrix_R, object, idxset)

  attr_list <- attributes(object)
  nr <- nrow(object)
  len <- sapply(attr_list, length)
  ind <- which(len == nr)
  if (length(ind) > 0) {
    nms <- names(attr_list)[ind]
    for (i in 1:length(ind)) {
      attr(ret, nms[i]) <- attr(object, nms[i])[idxset]
    }
  }
  return(structure(ret, class = "xgb.DMatrix"))
}
```

In lightgbm:
```
slice = function(idxset, ...) {
      
      # Perform slicing
      Dataset$new(NULL,
                  private$params,
                  self,
                  private$colnames,
                  private$categorical_feature,
                  private$predictor,
                  private$free_raw_data,
                  idxset,
                  NULL,
                  ...)
      
    }
```

This is a code for test the problem:

```
require(data.table)
require(Matrix)
require(lightgbm)

logloss <- function(actual, predicted, eps = 10^(-15)) {
  predicted <- pmin(pmax(predicted, eps), 1 - eps)
  -1 / length(actual) * (sum(actual * log(predicted) + (1 - actual) * log(1 - predicted)))
}

interested_only <- function(preds, dtrain) {
  observed <- getinfo(dtrain, 'label')
  interested <- attr(dtrain, 'interested')
  pred.dt <- data.table('interested' = interested, 'observed' = observed, 'preds' = preds)
  # Only interested in evaluate when interested == 1
  pred.dt <- pred.dt[interested == 1]
  ll <- logloss(pred.dt$observed, pred.dt$preds)
  return(list(metric = 'interested_only', value = ll))
}

# load in the agaricus dataset
data(agaricus.train, package = 'lightgbm')
dtrain_lgb <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label)

# We want train with all data but only evaluate in interested subset (which first column is 1)
# Xgboost uses attr for this
interested <- agaricus.train$data[, 1]
attr(dtrain_lgb, 'interested') <- interested

# lightgbm: attribute 'interested' isn't visible inside the custom evaluation function (problem with slice?)
param_lgb <- list(num_leaves = 32,
              learning_rate = 0.05,
              objective = 'binary')

lgb.cv(param_lgb, dtrain_lgb, nround = 20, nfold = 5, eval = interested_only)

# xgboost: attribute 'interested' is visible inside custom evaluation function
require(xgboost)
dtrain_xgb <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
attr(dtrain_xgb, 'interested') <- interested

param_xgb <- list(max_depth = 5,
              learning_rate = 0.05,
              objective = 'reg:logistic')
xgb.cv(param_xgb, dtrain_xgb, nround = 20, nfold = 5, eval_metric = interested_only)

```

**Comment by Laurae2** [comment]
@BlindApe there are issues currently with the `slice` method in LightGBM for attributes.

Related: https://github.com/Microsoft/LightGBM/issues/390

I wonder if there is a proper way to replicate attributes properly when slicing a `lgb.Dataset`.

**Comment by BlindApe** [comment]
Seems the same logic used in xgboost works:
`require(data.table)
require(Matrix)
require(lightgbm)

slice.lgb.Dataset <- function(dataset, idxset, ...) {
  
  ret <- dataset$slice(idxset, ...)
  attr_list <- attributes(dataset)
  nr <- nrow(dataset)
  len <- sapply(attr_list, length)
  ind <- which(len == nr)
  if (length(ind) > 0) {
    nms <- names(attr_list)[ind]
    for (i in 1:length(ind)) {
      attr(ret, nms[i]) <- attr(dataset, nms[i])[idxset]
    }
  }
  # Return sliced set
  invisible(ret) 
}

data(agaricus.train, package = 'lightgbm')
dtrain_lgb <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label)

interested <- agaricus.train$data[, 1]
attr(dtrain_lgb, 'interested') <- interested

folds <- sample(1:5, nrow(dtrain_lgb), replace = TRUE)
dtrain_lgb_1 <- slice.lgb.Dataset(dtrain_lgb, which(folds == 1))
names(attributes(dtrain_lgb_1))
attr(dtrain_lgb_1, 'interested')
`

**Comment by Laurae2** [comment]
@BlindApe you can make a PR to add it to LightGBM 📦 👍 

@guolinke What do you think about that slice method from @BlindApe ?

**Comment by guolinke** [comment]
@BlindApe @Laurae2 
Sure, welcome to create the PR. 

@Laurae2  BTW, is the slice according to group_id still hard ? we already fix it in python-package ?

**Comment by Laurae2** [comment]
@guolinke we need to recreate group ids when slicing.

Currently, group ids are per observation or the number of consecutive observations? (last time I tried I used consecutive observations, and had issues).

**Comment by guolinke** [comment]
@Laurae2 
Yeah, you can create an temporary group id (https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L241-L242) to preform split.

**Comment by Laurae2** [comment]
Will try to work on it, probably next weekend unless someone does it before me.

**Comment by BlindApe** [comment]
@Laurae2 I'll try to do it. I've seen that slice.lgb.Dataset isn't used in lgb.cv, but slice() instead.
I'm not sure if I change slice() with slice.lgb.Dataset() would be the best option.

**Comment by Laurae2** [comment]
@BlindApe What about creating a function named lgb.prepare.cv?  You can follow the naming I used to prepare datasets.

Input: a lgb.Dataset, a list of numeric vectors, and what to return (train folds? Test folds? Both?). Each vector: out of fold, so you have to check what the user wants initially from the function.

I'll comment on your PR when you start getting it done.

You might want to fix the slicing at the same time, and potentially the grouping issue.

**Comment by guolinke** [comment]
@BlindApe @Laurae2 any updates ?

**Comment by Laurae2** [comment]
@guolinke I'll try to work on it this weekend, it should fix #390 at the same time.

Do you have any R proper example for lambdarank like [here](https://github.com/Microsoft/LightGBM/issues/390#issue-220486691) with a dataset? I just want to make sure I am testing properly lambdarank.

**Comment by guolinke** [comment]
@Laurae2 you can use this(https://github.com/Microsoft/LightGBM/tree/master/examples/lambdarank).
I think this dataset can be add into R-package as well.
